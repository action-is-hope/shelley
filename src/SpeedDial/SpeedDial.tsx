import { forwardRef, ReactNode, useState } from "react";
import Add from "../icons/Add";
import { ButtonGroup } from "../ButtonGroup";
import { st, classes } from "./speedDial.st.css";
import { TooltipButton } from "../TooltipButton";

interface SpeedDialProps {
  children: ReactNode;
  tooltipTitle: string;
}

function SpeedDial(props: SpeedDialProps, ref: React.Ref<HTMLInputElement>) {
  const [speedDialOpen, triggerSpeedDial] = useState("false");
  const { children, tooltipTitle } = props;

  return (
    <div ref={ref} className={st(classes.root, { speedDialOpen })}>
      <TooltipButton
        data-id="tooltip-button"
        buttonProps={{
          onBlur: () => triggerSpeedDial("false"),
          onClick: () =>
            triggerSpeedDial(speedDialOpen === "false" ? "true" : "false"),
          icon: <Add />,
          className: classes.addButton,
        }}
        tooltip={tooltipTitle}
      />
      <ButtonGroup className={classes.buttonGroup}>{children}</ButtonGroup>
    </div>
  );
}

const _SpeedDial = forwardRef(SpeedDial);
export { _SpeedDial as SpeedDial };
