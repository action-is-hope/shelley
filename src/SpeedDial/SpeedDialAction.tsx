import { forwardRef, ReactElement } from "react";
import { TooltipButton } from "../Tooltip/TooltipButton";

interface SpeedDialActionProps {
  tooltipTitle: string;
  icon: ReactElement;
  onMouseDown: () => void;
}

function SpeedDialAction(
  props: SpeedDialActionProps,
  ref: React.Ref<HTMLInputElement>
) {
  const { tooltipTitle, icon, onMouseDown } = props;

  return (
    <div data-id="speed-dial-action" ref={ref}>
      <TooltipButton
        buttonProps={{
          onClick: onMouseDown,
          icon,
        }}
        tooltip={tooltipTitle}
      />
    </div>
  );
}

const _SpeedDialAction = forwardRef(SpeedDialAction);
export { _SpeedDialAction as SpeedDialAction };
