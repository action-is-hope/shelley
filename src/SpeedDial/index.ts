export * from "./SpeedDial";
export * from "./SpeedDialAction";
