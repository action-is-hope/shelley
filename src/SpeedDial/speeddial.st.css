@st-import [--spacing-unit] from "../styles/default/project.st.css";
@namespace "SpeedDial";

.root {
  -st-states:
    speedDialOpen(string);
  display: flex;
}

.root > .buttonGroup {
  margin-left: var(--spacing-unit);
}

.root:speedDialOpen(false) > div > .addButton > svg {
  transform: rotate(0);
  transition: transform 0.5s ease;
}

.root:speedDialOpen(true) > div > .addButton > svg {
  transform: rotate(45deg);
  transition: transform 0.5s ease;
}

.root:speedDialOpen(false) > .buttonGroup {
  visibility: hidden;
  opacity: 0;
  animation: slide-out .5s backwards;
  animation-fill-mode: backwards; 
}

.root:speedDialOpen(true) > .buttonGroup {
  visibility: visible;
  opacity: 1;
  animation: slide-in .5s forwards; 
}

@keyframes slide-in {
  0% {
    transform: translateX(-100%);
    opacity: 0;
    visibility: hidden;
  }

  100% {
    transform: translateX(0);
    opacity: 1;
    visibility: visible;
  }
}

@keyframes slide-out {
  0% {
    transform: translateX(0);
    opacity: 1;
    visibility: visible;
  }
  
  100% {
    transform: translateX(-100%);
    opacity: 0;
    visibility: hidden;
  }
}
