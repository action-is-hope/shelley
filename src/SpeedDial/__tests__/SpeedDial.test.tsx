import renderer from "react-test-renderer";
import { SpeedDial } from "../SpeedDial";

jest.mock("../speedDial.st.css", () => {
  const originalSt: object = jest.requireActual("../speedDial.st.css");

  return {
    ...originalSt,
    st: () => {
      return `speedDial___mock`;
    },
  } as object;
});

describe("SpeedDial", () => {
  it("renders correctly when closed", () => {
    const tooltipTitleMock = "Tooltip Test";
    const childrenMock = <span>Child Content</span>;

    const tree = renderer
      .create(
        <SpeedDial tooltipTitle={tooltipTitleMock}>{childrenMock}</SpeedDial>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
