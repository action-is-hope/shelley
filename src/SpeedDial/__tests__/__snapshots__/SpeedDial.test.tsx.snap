// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpeedDial renders correctly when closed 1`] = `
<div
  className="speedDial__Test__root"
  data-id="speed-dial"
>
  <div
    className="tooltipButton__Test__root"
    data-id="tooltip-button"
  >
    <button
      className="Button__Test__root Button__Test---tone-1-1 Button__Test---variant-7-primary Button__Test---vol-1-3 SpeedDial__Test__addButton"
      onBlur={[Function]}
      onClick={[Function]}
      onDragStart={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      type="button"
    >
       
      <svg
        aria-hidden={true}
        className="Icon__Test__root"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
      </svg>
    </button>
  </div>
  <div
    className="ButtonGroup__Test__root ButtonGroup__Test---orientation-10-horizontal SpeedDial1414287531__buttonGroup"
    data-id="button-group"
  >
    <span>
      Child Content
    </span>
  </div>
</div>
`;
